### PR TITLE
add script evaluation steps to bitcoin redeem and refund section

### PR DESCRIPTION
**Pull request by da-kami**
I know that for "a bitcoin programmer" this is all very clear... but I think it still helps to write it down - not everybody out there is a bitcoin programmer 😉

### CODE REVIEW

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 1]
```diff
@@ -143,6 +143,41 @@ For how to use this witness data to construct the redeem transaction see [BIP141
 To be notified of the redeem event, both parties MAY watch the blockchain for transactions that spend from the output and check that the witness data is in the above form.
 If Bitcoin is the `beta_ledger` (see [RFC003](./RFC-003-SWAP-Basic.md)), then the funder MUST watch for such a transaction and  extract the `secret` from its witness data and continue the protocol.
 
+#### Evaluation of the script for redeem
+
+**Given** a bitcoin script address that holds the following (hashed) information:
+1. `script code`
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: The wording is strange. `1.` contains `2.`, `3.` & `4.`.

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 2]
```diff
@@ -143,6 +143,41 @@ For how to use this witness data to construct the redeem transaction see [BIP141
...
+2. `hash of secret`
+3. `hash of redeem_pubkey`
+4. `hash of refund_pubkey`
+
+**When** a transaction is made to the address that pushes the following on the stack:
+1. redeem_signature
+2. redeem_pubkey
+3. secret
+4. byte (`01`)
+5. contract_script
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: The contract script is not pushed on the stack, it is evaluated.

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 3]
```diff
@@ -143,6 +143,41 @@ For how to use this witness data to construct the redeem transaction see [BIP141
...
+
+**Then** the script will be evaluated like this: 
+
+1. pop element of stack: script
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: I don't think this is how it's done, let's investigate together.

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 4]
```diff
@@ -143,6 +143,41 @@ For how to use this witness data to construct the redeem transaction see [BIP141
...
+    * pop element of stack: byte
+    * byte set to `01`, enter redeem block
+3. `    OP_SIZE 32 OP_EQUALVERIFY`: 
+    * peek element of stack: secret
+    * check size equals 32 bytes
+4. `    OP_SHA256 <secret_hash> OP_EQUALVERIFY`: 
+    * pop scret of stack and hash it
+    * match it against `hash of secret`
+5. `    OP_DUP OP_HASH160 <redeem_identity> OP_EQUALVERIFY OP_CHECKSIG`
+    * [standard evaluation of transaction to bitcoin address](https://en.bitcoin.it/wiki/Script#Standard_Transaction_to_Bitcoin_address_.28pay-to-pubkey-hash.29), address defined in transaction
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: Yes we use the same piece of code but hthis is not a "standard evaluation of transaction to bitcoin address". We never used the address related to this `redeem_identity` so I think it'd just confuse people if you write that.

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 5]
```diff
@@ -143,6 +143,41 @@ For how to use this witness data to construct the redeem transaction see [BIP141
...
+
+Note that the address to spend to can be any address, not necessarily the `redeem_pubkey`, any
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]:
```suggestion
Note that the address to spend to (ouput) can be any address, not necessarily the `redeem_pubkey`, any 
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 6]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
 
 To be notified of the refund event, both parties MAY watch the blockchain for transactions that spend from the output and check that the witness data is in the above form.
 
+#### Evaluation of the script for refund
+
+**Given** a bitcoin script address that holds the following (hashed) information:
+1. `script code`
+2. `hash of secret`
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: Well actually it's the same script for redeem and refund, the wording suggest it's 2 different scripts.

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 7]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
...
+3. `hash of redeem_pubkey`
+4. `hash of refund_pubkey`
+
+**When** a transaction is made to the address that pushes the following on the stack:
+1. refund_signature
+2. refund_pubkey
+3. byte (`01`)
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]:
```suggestion
3. byte (`00`)
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 8]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
...
+4. contract_script
+
+**Then** the script will be evaluated like this: 
+
+1. pop element of stack: script
+    * match script hash against address
+    * execute script
+2. `OP_IF`: 
+    * pop element of stack: byte
+    * byte set to `00`, enter refund block
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]:
```suggestion
    * byte set to `00`, enter `OP_ELSE`
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 9]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
...
+3. byte (`01`)
+4. contract_script
+
+**Then** the script will be evaluated like this: 
+
+1. pop element of stack: script
+    * match script hash against address
+    * execute script
+2. `OP_IF`: 
+    * pop element of stack: byte
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]:
```suggestion
    * evaluate and pop element of stack: byte
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 10]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
...
+    * byte set to `00`, enter refund block
+3. `    <expiry> OP_CHECKLOCKTIMEVERIFY OP_DROP`: 
+    * put expiry constant on stack
+    * check if expiry timestamp is in the past
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]:
```suggestion
    * check if expiry timestamp is in the past, if not, script execution fails
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 11]
```diff
@@ -157,6 +192,39 @@ The funder can use the following witness data to spend the output after the `exp
...
+    * remove expiry constant from stack
+4. `    OP_DUP OP_HASH160 <redeem_identity> OP_EQUALVERIFY OP_CHECKSIG`
+    * [standard evaluation of transaction to bitcoin address](https://en.bitcoin.it/wiki/Script#Standard_Transaction_to_Bitcoin_address_.28pay-to-pubkey-hash.29), address defined in transaction
```
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: Same comment than for redeem

**Comment by thomaseizinger** [comment]
> I know that for "a bitcoin programmer" this is all very clear... but I think it still helps to write it down - not everybody out there is a bitcoin programmer wink

I am not quite sure if adding a step by step description of the script is helpful in the RFC. It is a part that has to be maintained and kept in sync with the actual script (which already explains what is done but in "bitcoin speak").

As you said, for a person knowledgeable about the Bitcoin scripting language, this does not add any value. It is kind of the same discussion as with code comments. You should not duplicate _what_ is done in the code but rather explain _why_ if it is important.

I am against adding such an evaluation of the script.

What might be helpful though is an explanation, why the script contains the elements it contains:

Redeem:
1. hash of secret so that the receiver cannot just take the money
2. signature check for redeem address so that not everyone cannot take the money once they learn the secret

Refund:
1. timelock so that the it cannot be refunded early
2. signature check for the same reasons as above

Adding that + highlighting, which parts of the script are responsible for this functionality could maybe add some clarity to how the script works given basic understanding of stack based languages.

File: RFC-005-SWAP-Basic-Bitcoin.md (hunk 4 above)
Comment (da-kami) [reply RFC-005-SWAP-Basic-Bitcoin.md]: Happy to rename it - just named it like the example in the documentation: https://en.bitcoin.it/wiki/Script#Standard_Transaction_to_Bitcoin_address_.28pay-to-pubkey-hash.29
I agree it is probably misleading, will try to change it.

File: RFC-005-SWAP-Basic-Bitcoin.md (hunk 1 above)
Comment (da-kami) [reply RFC-005-SWAP-Basic-Bitcoin.md]: As far as I understood it, the transaction contains not only the parameters but also the script to be executed. The script's hash is checked against the address to evaluate if the script given is correct and will be executed. Is that incorrect?

**Comment by D4nte** [comment]
> Adding that + highlighting, which parts of the script are responsible for this functionality could maybe add some clarity to how the script works given basic understanding of stack based languages.

The issue is that today, if someone tries to understand COMIT, all they can read are the RFCs, hence, they do need *basic understanding of stack based languages*.
@da-kami is trying to fill this gap with this graph. An alternative would be much more time consuming by creating "user documentation". I support @da-kami in this initiative, especially that we are unlikely to change the Bitcoin script.

**Comment by da-kami** [comment]
> What might be helpful though is an explanation, why the script contains the elements it contains: 
> ...

This could be achieved by adding comments to the script code similar to what we do in the Ethereum contract assembly (https://github.com/comit-network/RFCs/blob/master/RFC-007-SWAP-Basic-Ether.md). 
However, I think it would be helpful to additionally step through the evaluation as, like @D4nte pointed out, this is the only documentation we have at the moment.

**Comment by da-kami** [comment]
I removed the separate sections but added a version of the contract that describes the execution in rough terms. This is similar to the comments for the Ethereum contract. 

@thomaseizinger can you live with that?
@D4nte please have another quick look

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 12]
```diff
@@ -106,6 +106,62 @@ To compute the exact Bitcoin script bytes of the contract, implementations shoul
 | `refund_identity` | 74                     | 93                    | 20             | See [Parameters](#parameters)                                              |
 | `6888ac`          | 94                     | 96                    | 3              | `OP_ENDIF OP_EQUALVERIFY OP_CHECKSIG`                                      |
 
+
+Script execution overview:
+
+```
+// cur element on stack: byte
```
Comment (bonomat) [inline RFC-005-SWAP-Basic-Bitcoin.md]: if you want to be precise, you would need to mention all elements on the stack, e.g. current stack
```
<signature>
<pubkey>
<secret>
<01>
```

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 13]
```diff
@@ -106,6 +106,62 @@ To compute the exact Bitcoin script bytes of the contract, implementations shoul
...
+//   evaluate byte: 01
+//   OP_IF (redeem) is entered
+//   pop byte
```
Comment (bonomat) [inline RFC-005-SWAP-Basic-Bitcoin.md]: I'm not sure if this description makes sense in that way: what does `evaluate byte: 01` mean? 
`OP_IF` pops an element from the stack and if this is != 0 it continues with the branch.

File: RFC-005-SWAP-Basic-Bitcoin.md (hunk 12 above)
Comment (bonomat) [inline RFC-005-SWAP-Basic-Bitcoin.md]: You probably want to say: `top stack element: byte` but then again, which byte?

File: RFC-005-SWAP-Basic-Bitcoin.md [hunk 14]
```diff
@@ -106,6 +106,62 @@ To compute the exact Bitcoin script bytes of the contract, implementations shoul
...
+OP_IF
+
+    // cur element on stack: secret
+    //   peek secret size and push size on stack
+    //   push constant 32 on stack
+    //   pop 2 element of stack (32, secret size) and check equality
+    OP_SIZE 32 OP_EQUALVERIFY
```
Comment (bonomat) [inline RFC-005-SWAP-Basic-Bitcoin.md]: Not 100% correct: `OP_EQUALVERIFY` is a combination of `OP_EQUAL` and `OP_VERIFY` meaning under the hood it first executes equal (which consumes the top 2 elements) pushes back the result (0 or 1) and then verifies if the top of the stack != 0.

File: RFC-005-SWAP-Basic-Bitcoin.md (hunk 12 above)
Comment (D4nte) [inline RFC-005-SWAP-Basic-Bitcoin.md]: This is the top element on the stack, current elements on stack are:
byte
secret
pubkey
signature

**Comment by da-kami** [comment]
After a discussion with @bonomat the decision is to describe these details in a blogpost (on the CoBloX blog) rather than in the RFC. Thus, this PR will be closed.

**Comment by da-kami** [comment]
After a discussion with @bonomat the decision is to describe these details in a blogpost (on the CoBloX blog) rather than in the RFC. Thus, this PR will be closed.

**Pull request closed without merging**